dashbuilder-client/dashbuilder-navigation-client/src/main/java/org/dashbuilder/client/navigation/widget/NavTreeWidgetView.java

```diff
@@ -0,0 +1,126 @@
+import com.google.gwt.user.client.Event;
+import com.google.gwt.user.client.ui.IsWidget;
+import org.dashbuilder.common.client.widgets.AlertBox;
+import org.jboss.errai.common.client.dom.Div;
+import org.jboss.errai.common.client.dom.Node;
+import org.jboss.errai.ui.shared.api.annotations.DataField;
+import org.jboss.errai.ui.shared.api.annotations.Templated;
+import org.uberfire.mvp.Command;
+
+@Dependent
+@Templated
+public class NavTreeWidgetView extends TargetDivNavWidgetView<NavTreeWidget>
+    implements NavTreeWidget.View {
+
+    @Inject
+    @DataField
+    Div mainDiv;
+
+    NavTreeWidget presenter;
+
+    @Inject
+    public NavTreeWidgetView(AlertBox alertBox) {
+        super(alertBox);
+    }
+
+    @Override
+    public void init(NavTreeWidget presenter) {
+        this.presenter = presenter;
+        super.navWidget = mainDiv;
+    }
+
+    @Override
+    public void addDivider() {
+        // Useless in a tree
+    }
+
+    @Override
+    public void setLevel(int level) {
+        int margin = level*10;
+        mainDiv.getStyle().setProperty("margin-left", margin + "px");
+        if (level == 0) {
+            mainDiv.getStyle().setProperty("padding", "10px");
+        }
+    }
+
+    @Override
+    public void addRuntimePerspective(String id, String name, String description, Command onClicked) {
+        this.addItem("pficon-virtual-machine", id, name, description, onClicked);
+    }
+
+    @Override
+    public void addPerspective(String id, String name, String description, Command onClicked) {
+        this.addItem("pficon-screen", id, name, description, onClicked);
+    }
+
+    @Override
+    public void addGroupItem(String id, String name, String description, IsWidget widget) {
+        this.addItem("pficon-folder-open",id, name, description, null);
+        super.addGroupItem(id, name, description, widget);
+    }
+
+    @Override
+    protected void setSelectedEnabled(boolean enabled) {
+        String cname = selectedItem.getClassName();
+        if (!cname.equals("navtree-widget-non-clickable")) {
+            if (enabled) {
+                selectedItem.setClassName("navtree-widget-clicked");
+            } else {
+                selectedItem.setClassName("navtree-widget-non-clicked");
+            }
+        }
+    }
+
+    protected void addItem(String iconClass, String id, String name, String description, Command onClicked) {
+        Element nameEl = onClicked != null ? Document.get().createAnchorElement() : Document.get().createSpanElement();
+        nameEl.setInnerText(name);
+        nameEl.setClassName(onClicked != null ? "navtree-widget-non-clicked" : "navtree-widget-non-clickable");
+        if (description != null && !description.equals(name)) {
+            nameEl.setTitle(description);
+        }
+
+        SpanElement iconSpan = Document.get().createSpanElement();
+        iconSpan.setClassName("navtree-widget-icon " + iconClass);
+
+        DivElement div = Document.get().createDivElement();
```

Comment:
For this type of inner component, like I saw some of them at other places, I would recommend creating Errai Widgets to encapsulate the logic.

Reply:
Do you mean creating another UI component for representing single tree items?

Reply:
Yes, I personally prefer to use smaller errai components and instantiate them.